### PR TITLE
light: make fraction parts uint64, ensuring that it is always positive

### PR DESCRIPTION
**Pull request by cmwaters**
## Description

This is a minor change to ensure that only positive fractions can be created (tendermint only uses fractions for trust level which we need to be positive).

Alternatively I can just add non negative checks to the validate function but I think we are looking at going down the unsigned path anyway.

Closes: #5651

### CODE REVIEW

File: libs/math/fraction.go [hunk 1]
```diff
@@ -38,5 +37,12 @@ func ParseFraction(f string) (Fraction, error) {
 	if err != nil {
 		return Fraction{}, fmt.Errorf("incorrect formatting, err: %w", err)
 	}
-	return Fraction{Numerator: numerator, Denominator: denominator}, nil
+	if denominator == 0 {
+		return Fraction{}, errors.New("denominator can't be 0")
+	}
+	if numerator < 0 || denominator < 0 {
+		return Fraction{}, fmt.Errorf("numerator and denominator can't be negative. Got %d and %d respectively",
```
Comment (alexanderbez) [inline libs/math/fraction.go:44]:
```suggestion
		return Fraction{}, fmt.Errorf("numerator (%s) and denominator (%s) cannot be negative",
```

**Review by alexanderbez — APPROVED** [review]
This is a breaking change it seems. Do we need a changelog entry?

File: libs/math/fraction.go [hunk 2]
```diff
@@ -38,5 +37,12 @@ func ParseFraction(f string) (Fraction, error) {
 	if err != nil {
 		return Fraction{}, fmt.Errorf("incorrect formatting, err: %w", err)
 	}
-	return Fraction{Numerator: numerator, Denominator: denominator}, nil
+	if denominator == 0 {
+		return Fraction{}, errors.New("denominator can't be 0")
+	}
+	if numerator < 0 || denominator < 0 {
```
Comment (melekes) [inline libs/math/fraction.go:43]: we can use strconv.ParseUint functions instead

Comment (melekes) [reply libs/math/fraction.go:43]: should we also check they are not greater than `math.MaxInt64` to prevent possible overflow?

**Review by melekes — CHANGES_REQUESTED** [review]

File: libs/math/fraction.go (hunk 2 above)
Comment (cmwaters) [reply libs/math/fraction.go:43]: I think, looking at the code, that strconv.ParseUint should also ensure that it is less than MaxInt64

Comment (melekes) [reply libs/math/fraction.go]: uint64 and int64 have different max values, so no

Comment (cmwaters) [reply libs/math/fraction.go]: ah you mean in the case that we convert it. Yeah that makes sense

**Comment by melekes** [comment]
should we remove https://github.com/tendermint/tendermint/pull/5655/files#diff-c4a5571b309cf76086cc8f6614c957ba58f9abec7ada7088fc74dd1d2a5d7d3cL778 and update ValidateTrustLevel function?

**Comment by cmwaters** [comment]
Can you elaborate on the part we should remove? The link isn't specifically showing me any part of the code. 

Also what changes do we need to make to ValidateTrustLevel?

File: CHANGELOG_PENDING.md [hunk 3]
```diff
@@ -23,6 +23,8 @@ Friendly reminder, we have a [bug bounty program](https://hackerone.com/tendermi
 
 ### FEATURES
 
+- [libs/math] \#5665 Make fractions unsigned integers
```
Comment (marbar3778) [inline CHANGELOG_PENDING.md:26]:
```suggestion
- [libs/math] \#5665 Make fractions unsigned integers (uint64) (@cmwaters)
```
Not sure this is a feature. May be better under Improvements

Comment (cmwaters) [reply CHANGELOG_PENDING.md:26]: Yeah I had this small mental debate because it's not really an improvement either

Comment (cmwaters) [reply CHANGELOG_PENDING.md:26]: fractions can be negative - it's just it makes sense in the context of Tendermint that fractions are positive only

**Comment by melekes** [comment]
https://github.com/tendermint/tendermint/blob/8aa47c7da5b854a69f367f316a31c03d2fc1e942/types/validator_set.go#L778-L780

actually, nevermind. we still need these checks in both https://github.com/tendermint/tendermint/blob/8aa47c7da5b854a69f367f316a31c03d2fc1e942/types/validator_set.go#L778-L780 and ValidateTrustLevel.

**Review by melekes — APPROVED** [review]
👍

**Pull request merged**